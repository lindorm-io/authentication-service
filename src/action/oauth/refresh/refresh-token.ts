import Joi from "@hapi/joi";
import { IAuthContext, ICreateTokensData } from "../../../typing";
import { JOI_GRANT_TYPE } from "../../../constant";
import { createTokens, extendSession } from "../../../support";
import { Permission } from "../../../enum";
import { InvalidPermissionError, InvalidRefreshTokenError } from "../../../error";

export interface IPerformRefreshTokenOptions {
  grantType: string;
  responseType: string;
}

const schema = Joi.object({
  grantType: JOI_GRANT_TYPE,
  responseType: Joi.string().required(),
});

export const performRefreshToken = (ctx: IAuthContext) => async (
  options: IPerformRefreshTokenOptions,
): Promise<ICreateTokensData> => {
  await schema.validateAsync(options);

  const { client, repository, token } = ctx;
  const { responseType } = options;
  const {
    refresh: { id: refreshId, authMethodsReference, permission },
  } = token;

  const session = await extendSession(ctx)();
  const account = await repository.account.find({ id: session.accountId });

  if (account.permission === Permission.LOCKED) {
    throw new InvalidPermissionError();
  }

  if (account.permission !== permission) {
    throw new InvalidRefreshTokenError(refreshId);
  }

  return createTokens(ctx)({
    account,
    authMethodsReference,
    client,
    responseType,
    session,
  });
};
