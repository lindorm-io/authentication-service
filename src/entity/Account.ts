import { AccountEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { Permission } from "@lindorm-io/jwt";

export interface IAccountOTP {
  signature: string;
  uri: string;
}

export interface IAccountPassword {
  signature: string;
  updated: Date;
}

export interface IAccount extends IEntity {
  email: string;
  identityLinked: boolean;
  otp: IAccountOTP;
  password: IAccountPassword;
  permission: string;
}

export interface IAccountOptions extends IEntityBaseOptions {
  email: string;
  identityLinked?: boolean;
  otp?: IAccountOTP;
  password?: IAccountPassword;
  permission?: string;
}

export class Account extends EntityBase implements IAccount {
  private _email: string;
  private _identityLinked: boolean;
  private _otp: IAccountOTP;
  private _password: IAccountPassword;
  private _permission: string;

  constructor(options: IAccountOptions) {
    super(options);
    this._email = options.email.toLowerCase();
    this._identityLinked = options.identityLinked || false;
    this._otp = {
      signature: options.otp?.signature || null,
      uri: options.otp?.uri || null,
    };
    this._password = {
      signature: options.password?.signature || null,
      updated: options.password?.updated || null,
    };
    this._permission = options.permission || Permission.USER;
  }

  public get email(): string {
    return this._email;
  }
  public set email(email: string) {
    this._email = email.toLowerCase();
    this.addEvent(AccountEvent.EMAIL_CHANGED, { email: this._email });
  }

  public get identityLinked(): boolean {
    return this._identityLinked;
  }
  public set identityLinked(identityLinked: boolean) {
    this._identityLinked = identityLinked;
    this.addEvent(AccountEvent.EMAIL_CHANGED, { identityLinked: this._identityLinked });
  }

  public get otp(): IAccountOTP {
    return this._otp;
  }
  public set otp(otp: IAccountOTP) {
    this._otp = otp;
    this.addEvent(AccountEvent.OTP_CHANGED, { otp: this._otp });
  }

  public get password(): IAccountPassword {
    return this._password;
  }
  public set password(password: IAccountPassword) {
    this._password = password;
    this.addEvent(AccountEvent.PASSWORD_CHANGED, { password: this._password });
  }

  public get permission(): string {
    return this._permission;
  }
  public set permission(permission: string) {
    this._permission = permission;
    this.addEvent(AccountEvent.PERMISSION_CHANGED, { permission: this._permission });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== AccountEvent.CREATED) continue;
      throw new EntityCreationError("Account");
    }
    this.addEvent(AccountEvent.CREATED, {
      email: this._email,
      permission: this._permission,
      password: this._password,
      otp: this._otp,
      created: this._created,
      updated: this._updated,
    });
  }
}
