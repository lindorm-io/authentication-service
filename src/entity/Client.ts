import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { ClientEvent } from "../enum";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  emailAuthorizationUri: string;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  emailAuthorizationUri?: string;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _emailAuthorizationUri: string;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._emailAuthorizationUri = options?.emailAuthorizationUri || null;
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get emailAuthorizationUri(): string {
    return this._emailAuthorizationUri;
  }
  public set emailAuthorizationUri(uri: string) {
    this._emailAuthorizationUri = uri;
    this.addEvent(ClientEvent.EMAIL_AUTHORIZATION_URI_CHANGED, { uri: this._emailAuthorizationUri });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new Error("Client has already been created");
    }
    this.addEvent(ClientEvent.CREATED, {
      name: this._name,
      emailAuthorizationUri: this._emailAuthorizationUri,
      approved: this._approved,
      description: this._description,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
