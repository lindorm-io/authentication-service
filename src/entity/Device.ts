import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { DeviceEvent } from "../enum";

export interface IDevicePIN {
  signature: string;
  updated: Date;
}

export interface IDevice extends IEntity {
  accountId: string;
  name: string;
  pin: IDevicePIN;
  publicKey: string;
  secret: string;
}

export interface IDeviceOptions extends IEntityBaseOptions {
  accountId: string;
  name?: string;
  pin?: IDevicePIN;
  publicKey: string;
  secret?: string;
}

export class Device extends EntityBase implements IDevice {
  private _accountId: string;
  private _name: string;
  private _pin: IDevicePIN;
  private _publicKey: string;
  private _secret: string;

  constructor(options: IDeviceOptions) {
    super(options);
    this._accountId = options.accountId;
    this._name = options.name || null;
    this._pin = {
      signature: options.pin?.signature || null,
      updated: options.pin?.updated || null,
    };
    this._publicKey = options.publicKey;
    this._secret = options.secret || null;
  }

  public get accountId(): string {
    return this._accountId;
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(DeviceEvent.NAME_CHANGED, { name: this._name });
  }

  public get pin(): IDevicePIN {
    return this._pin;
  }
  public set pin(pin: IDevicePIN) {
    this._pin = pin;
    this.addEvent(DeviceEvent.PIN_CHANGED, { pin: this._pin });
  }

  public get publicKey(): string {
    return this._publicKey;
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(DeviceEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== DeviceEvent.CREATED) continue;
      throw new Error("Device has already been created");
    }
    this.addEvent(DeviceEvent.CREATED, {
      accountId: this._accountId,
      name: this._name,
      pin: this._pin,
      publicKey: this._publicKey,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
