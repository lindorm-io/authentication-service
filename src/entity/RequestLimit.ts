import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { GrantType } from "../enum";

export interface IRequestLimit extends IEntity {
  backOffUntil: Date;
  failedTries: number;
  grantType: GrantType;
  subject: string;
}

export interface IRequestLimitOptions extends IEntityBaseOptions {
  backOffUntil?: Date;
  failedTries?: number;
  grantType: GrantType;
  subject: string;
}

export class RequestLimit extends EntityBase implements IRequestLimit {
  private _backOffUntil: Date;
  private _failedTries: number;
  private _grantType: GrantType;
  private _subject: string;

  constructor(options: IRequestLimitOptions) {
    super(options);
    this._backOffUntil = options.backOffUntil || null;
    this._failedTries = options.failedTries || 1;
    this._grantType = options.grantType;
    this._subject = options.subject;
  }

  public get backOffUntil(): Date {
    return this._backOffUntil;
  }
  public set backOffUntil(backOffUntil: Date) {
    this._backOffUntil = backOffUntil;
    this._updated = new Date();
  }

  public get failedTries(): number {
    return this._failedTries;
  }
  public set failedTries(failedTries: number) {
    this._failedTries = failedTries;
    this._updated = new Date();
  }

  public get grantType(): GrantType {
    return this._grantType;
  }

  public get subject(): string {
    return this._subject;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
