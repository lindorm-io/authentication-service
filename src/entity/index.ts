export * from "./Account";
export * from "./RequestLimit";
export * from "./Session";
