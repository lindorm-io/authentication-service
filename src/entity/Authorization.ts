import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { GrantType } from "../enum";
import { Scope } from "@lindorm-io/jwt";

export interface IAuthorization extends IEntity {
  challengeId: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export interface IAuthorizationOptions extends IEntityBaseOptions {
  challengeId?: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId?: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode?: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export class Authorization extends EntityBase implements IAuthorization {
  readonly challengeId: string;
  readonly clientId: string;
  readonly codeChallenge: string;
  readonly codeMethod: string;
  readonly deviceId: string;
  readonly email: string;
  readonly expires: Date;
  readonly grantType: GrantType;
  readonly otpCode: string;
  readonly redirectUri: string;
  readonly responseType: string;
  readonly scope: Array<Scope>;

  constructor(options: IAuthorizationOptions) {
    super(options);

    this.challengeId = options.challengeId || null;
    this.clientId = options.clientId;
    this.codeChallenge = options.codeChallenge;
    this.codeMethod = options.codeMethod;
    this.deviceId = options.deviceId || null;
    this.email = options.email;
    this.expires = options.expires;
    this.grantType = options.grantType;
    this.otpCode = options.otpCode || null;
    this.redirectUri = options.redirectUri;
    this.responseType = options.responseType;
    this.scope = options.scope;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
