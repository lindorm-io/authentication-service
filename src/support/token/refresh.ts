import { Account, Device, Session } from "../../entity";
import { Audience } from "../../enum";
import { Client } from "@lindorm-io/koa-client";
import { IAuthContext } from "../../typing";
import { ITokenIssuerSignData } from "@lindorm-io/jwt";

export interface IGetRefreshTokenOptions {
  account: Account;
  authMethodsReference: string;
  client: Client;
  device?: Device;
  scope: string;
  session: Session;
}

export const getRefreshToken = (ctx: IAuthContext) => (options: IGetRefreshTokenOptions): ITokenIssuerSignData => {
  const { logger, issuer } = ctx;
  const { tokenIssuer } = issuer;
  const { authMethodsReference, client, device, account, scope, session } = options;

  logger.debug("creating refresh token", {
    account: account.id,
    client: client.id,
    device: device?.id,
    scope,
    session: session.id,
  });

  return tokenIssuer.sign({
    audience: Audience.REFRESH,
    authMethodsReference: authMethodsReference,
    clientId: client.id,
    deviceId: device?.id,
    expiry: session.expires,
    id: session.refreshId,
    permission: account.permission,
    scope,
    subject: session.id,
  });
};
