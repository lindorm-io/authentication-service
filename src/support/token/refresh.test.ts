import MockDate from "mockdate";
import { Account, Session } from "../../entity";
import { Client } from "@lindorm-io/koa-client";
import { getRefreshToken } from "./refresh";
import { getTestAccount, getTestClient, getTestIssuer, getTestSession, logger } from "../../test";

jest.mock("jsonwebtoken", () => ({
  sign: (data: any) => data,
}));
jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("getRefreshToken", () => {
  let ctx: any;
  let account: Account;
  let client: Client;
  let session: Session;

  beforeEach(() => {
    ctx = {
      logger,
      issuer: { tokenIssuer: getTestIssuer() },
      metadata: { deviceId: "deviceId" },
    };

    account = getTestAccount("email@lindorm.io");
    client = getTestClient();
    session = getTestSession(account, client, "codeChallenge", "codeMethod");
  });

  test("should return a refresh token", () => {
    expect(
      getRefreshToken(ctx)({
        account,
        authMethodsReference: "authMethodsReference",
        client,
        scope: "scope",
        session,
      }),
    ).toMatchSnapshot();
  });
});
