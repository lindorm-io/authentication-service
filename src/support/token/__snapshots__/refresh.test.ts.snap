// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRefreshToken should return a refresh token 1`] = `
Object {
  "expires": 4070908800,
  "expiresIn": 2493046800,
  "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "level": undefined,
  "token": Object {
    "amr": "authMethodsReference",
    "aud": "refresh",
    "cid": "9ba5e0c5-f2aa-4706-a27d-66369b547b0b",
    "did": "deviceId",
    "exp": 4070908800,
    "iam": "user",
    "iat": 1577862000,
    "iss": "https://test.authentication.lindorm.io",
    "jti": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "nbf": 1577862000,
    "sco": "scope",
    "sub": "be3a62d1-24a0-401c-96dd-3aff95356811",
  },
}
`;
